File: src/geometry.js
/**
 * Pure point/segment geometry shared across the Voronoi pipeline.
 */

/**
 * Find L1 distance
 *
 * @param {Array} P1 - [x,y]
 * @param {Array} P2 - [x,y]
 * @returns {number}
 */
export function distance(P1, P2){
    return Math.abs(P1[0] - P2[0]) + Math.abs(P1[1] - P2[1]);
}

/**
 * Determine if two points are the same point
 *
 * @param {Array} P1 - [x,y]
 * @param {Array} P2 - [x,y]
 */
export function samePoint(P1, P2){
    return P1[0] === P2[0] && P1[1] === P2[1];
}

/**
 * Angle of vector P2-P1 in [0, 2*PI)
 *
 * @param {Array} P1 - [x,y]
 * @param {Array} P2 - [x,y]
 */
export function angle(P1, P2){
    let result = Math.atan2(P2[1] - P1[1], P2[0] - P1[0]);

    if(result < 0){
        result = Math.PI + Math.PI + result;
    }

    return result;
}

/**
 * Find intersection of two line segments, if it exists.
 * Returns null when the segments do not intersect.
 *
 * @param {LineSegment} L1 - [[x,y],[x,y]]
 * @param {LineSegment} L2 - [[x,y],[x,y]]
 * @returns {Array|null}
 */
export function segmentIntersection(L1, L2){

    var ua, ub, denom = (L2[1][1] - L2[0][1])*(L1[1][0] - L1[0][0]) - (L2[1][0] - L2[0][0])*(L1[1][1] - L1[0][1]);

    if (denom == 0) {
        return null;
    }
    ua = ((L2[1][0] - L2[0][0])*(L1[0][1] - L2[0][1]) - (L2[1][1] - L2[0][1])*(L1[0][0] - L2[0][0]))/denom;
    ub = ((L1[1][0] - L1[0][0])*(L1[0][1] - L2[0][1]) - (L1[1][1] - L1[0][1])*(L1[0][0] - L2[0][0]))/denom;

    if(
        !(ua >= 0 && ua <= 1 &&
        ub >= 0 && ub <= 1)
    ){
        return null;
    }

    return [
        L1[0][0] + ua*(L1[1][0] - L1[0][0]),
        L1[0][1] + ua*(L1[1][1] - L1[0][1])
    ];

}

File: py_ai/geometry.py
"""
Pure point/segment geometry shared across the Voronoi pipeline.
"""

import math


def angle(p1, p2):
    """Calculate angle between two points"""
    ang = math.atan2(p2[1] - p1[1], p2[0] - p1[0])
    if ang < 0:
        ang = math.pi + math.pi + ang
    return ang


def distance(p1, p2):
    """L1 distance between two points"""
    return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])


def same_point(p1, p2):
    """Check if two points are the same"""
    return p1[0] == p2[0] and p1[1] == p2[1]


def segment_intersection(l1, l2):
    """Find intersection of two line segments"""
    x0, y0 = l1[0]
    x1, y1 = l1[1]
    x2, y2 = l2[0]
    x3, y3 = l2[1]

    denom = (y3 - y2) * (x1 - x0) - (x3 - x2) * (y1 - y0)

    if denom == 0:
        return None

    ua = ((x3 - x2) * (y0 - y2) - (y3 - y2) * (x0 - x2)) / denom
    ub = ((x1 - x0) * (y0 - y2) - (y1 - y0) * (x0 - x2)) / denom

    if not (ua >= 0 and ua <= 1 and ub >= 0 and ub <= 1):
        return False

    return [x0 + ua * (x1 - x0), y0 + ua * (y1 - y0)]
